src/aws-cpp-sdk-core/source/utils/threading/DefaultExecutor.cpp

```diff
@@ -3,80 +3,187 @@
  * SPDX-License-Identifier: Apache-2.0.
  */
 
+#include <aws/core/utils/logging/LogMacros.h>
 #include <aws/core/utils/threading/DefaultExecutor.h>
 #include <aws/core/utils/threading/ThreadTask.h>
 
 #include <cassert>
 
-using namespace Aws::Utils::Threading;
-
-bool DefaultExecutor::SubmitToThread(std::function<void()>&&  fx)
-{
-    // Generalized lambda capture is C++14, using std::bind as a workaround to force moving fx (instead of copying)
-    std::function<void()> main = std::bind(
-            [this](std::function<void()>& storedFx)
-            {
-                storedFx();
-                Detach(std::this_thread::get_id());
-            },
-            std::move(fx)
-        );
-
-    State expected;
-    do
-    {
-        expected = State::Free;
-        if(m_state.compare_exchange_strong(expected, State::Locked))
-        {
-            std::thread t(std::move(main));
-            const auto id = t.get_id(); // copy the id before we std::move the thread
-            m_threads.emplace(id, std::move(t));
-            m_state = State::Free;
-            return true;
-        }
+namespace Aws {
+namespace Utils {
+namespace Threading {
+
+static const char DEFAULT_EXECUTOR_LOG_TAG[] = "DefaultExecutor";
+
+class DefaultExecutorTask;
+using DefaultExecutorTaskPair = std::pair<std::thread, DefaultExecutorTask*>;
+
+struct DefaultExecutor::impl {
+  impl() = default;
+  impl(const impl&) = delete;
+  impl(impl&&) = delete;
+  impl& operator=(const impl&) = delete;
+  impl& operator=(impl&&) = delete;
+
+  enum class State { Free, Locked, Shutdown };
+
+  ~impl();
+
+  bool SubmitToThread(const std::shared_ptr<impl>& spThis, std::function<void()>&& fx);
+  void WaitUntilStopped();
+  void Detach(std::thread::id id);
+
+ private:
+  std::atomic<State> m_state{State::Free};
+  Aws::UnorderedMap<std::thread::id, DefaultExecutorTaskPair> m_tasks;
+};
+
+class DefaultExecutorTask {
+ public:
+  DefaultExecutorTask(std::function<void()>&& task, const std::shared_ptr<DefaultExecutor::impl>& pExecutor)
+      : m_task(std::move(task)), m_executor(pExecutor) {
+    assert(m_task);
+  }
+  DefaultExecutorTask(const DefaultExecutorTask&) = delete;
+  DefaultExecutorTask& operator=(const DefaultExecutorTask&) = delete;
+  DefaultExecutorTask(DefaultExecutorTask&&) = default;
+
+  /**
+   * Detaches the task from the executor
+   */
+  void DoNotDetach() { m_executor.reset(); }
+  /**
+   * Starts task execution on a new thread
+   */
+  static DefaultExecutorTaskPair Launch(DefaultExecutorTask* task) {
+    if (!task) {
+      AWS_LOGSTREAM_FATAL(DEFAULT_EXECUTOR_LOG_TAG, "Attempted to launch an empty task");
+      return {};
+    }
+    return {std::thread(&DefaultExecutorTask::Execute, task), task};
+  }
+
+ private:
+  void Execute() {
+    assert(m_task);
+    m_task();
+    if (const auto spExecutor = m_executor.lock()) {
```

Comment:
if m_executor pimpl is being destructed, the lock() will return null
and `Aws::Delete(this)` will result in Detach being skipped, then this task is deleted

then the destructor of pimpl may try to call because the task is still present on the task map.
```
taskItem.second.second->DoNotDetach();
```